### PR TITLE
Schedule backfills in the scheduler

### PR DESCRIPTION
**Pull request by dstandish**
(cherry picked from commit 2a581abc99988ea9d3efa67311d04716696ccca4)

### CODE REVIEW

File: airflow/migrations/versions/0036_3_0_0_add_backfill_to_dag_run_model.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def downgrade():
+    """Unapply Add backfill to dag run model."""
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table("dag_run", schema=None) as batch_op:
+        batch_op.drop_constraint(batch_op.f("dag_run_backfill_id_fkey"), type_="foreignkey")
+        batch_op.drop_column("backfill_id")
+
+    # ### end Alembic commands ###
```
Comment (jedcunningham) [inline airflow/migrations/versions/0036_3_0_0_add_backfill_to_dag_run_model.py]:
```suggestion
    """Apply Add backfill to dag run model."""
    with op.batch_alter_table("dag_run", schema=None) as batch_op:
        batch_op.add_column(sa.Column("backfill_id", sa.Integer(), nullable=True))
        batch_op.create_foreign_key(
            batch_op.f("dag_run_backfill_id_fkey"), "backfill", ["backfill_id"], ["id"]
        )


def downgrade():
    """Unapply Add backfill to dag run model."""
    with op.batch_alter_table("dag_run", schema=None) as batch_op:
        batch_op.drop_constraint(batch_op.f("dag_run_backfill_id_fkey"), type_="foreignkey")
        batch_op.drop_column("backfill_id")
```

nit

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -4769,6 +4773,375 @@ def test_max_active_runs_in_a_dag_doesnt_stop_running_dagruns_in_otherdags(self,
         assert dag1_running_count == 1
         assert running_count == 11
 
+    def test_dag1_is_mutated(self, dag_maker, session):
```
Comment (jedcunningham) [inline tests/jobs/test_scheduler_job.py]: 1) this seems odd - something specific to dag_maker? 2) did you mean to keep this in here?

File: airflow/jobs/scheduler_job_runner.py [hunk 3]
```diff
@@ -1548,33 +1555,51 @@ def _update_state(dag: DAG, dag_run: DagRun):
...
-                    span.add_event(
-                        name="dag_run",
-                        attributes={
-                            "run_id": dag_run.run_id,
-                            "dag_id": dag_run.dag_id,
-                            "conf": str(dag_run.conf),
-                        },
+            active_runs = active_runs_of_dags[(dag_id, backfill_id)]
+            if backfill_id is not None:
+                if active_runs >= backfill.max_active_runs:
```
Comment (jedcunningham) [inline airflow/jobs/scheduler_job_runner.py:1568]: This means in practice we have a totally separate max_active_runs for scheduled and each backfill?

Intuitively, it feels like the DAGs max_active_runs should be honored across types.

File: tests/jobs/test_scheduler_job.py [hunk 4]
```diff
@@ -4769,6 +4773,375 @@ def test_max_active_runs_in_a_dag_doesnt_stop_running_dagruns_in_otherdags(self,
...
+        scheduler_job = Job()
+        self.job_runner = SchedulerJobRunner(job=scheduler_job, subdir=os.devnull)
+
+        scheduler_job.executor = MockExecutor(do_update=False)
+        self.job_runner.processor_agent = mock.MagicMock(spec=DagFileProcessorAgent)
+
+        self.job_runner._start_queued_dagruns(session)
+        session.flush()
+        self.job_runner._start_queued_dagruns(session)
+        session.flush()
```
Comment (jedcunningham) [inline tests/jobs/test_scheduler_job.py:4810]: Why do we need to do it twice? Just the volume of runs?

File: airflow/jobs/scheduler_job_runner.py (hunk 3 above)
Comment (dstandish) [reply airflow/jobs/scheduler_job_runner.py:1568]: yeah, tricky to decide the right interface. the idea with doing it separately is you have complete control over both kinds of runs.

if they are mixed in together, the number of concurrent backfill dag runs depends on how many prod runs are running.  backfill runs could get starved out by prod runs (at least in theory) -- not likely i'd bet.

File: tests/jobs/test_scheduler_job.py (hunk 2 above)
Comment (dstandish) [reply tests/jobs/test_scheduler_job.py]: yeah i didn't really mean to include it. was just going to post on slack... "isn't this odd"

File: tests/jobs/test_scheduler_job.py (hunk 4 above)
Comment (dstandish) [reply tests/jobs/test_scheduler_job.py:4810]: yeah i think so. based this test off of one from ephraim.  but pretty sure that's the reason.  there are 30+ runs created and max runs default is 20

**Review by jedcunningham — APPROVED** [review]

**Comment by jscheffl** [comment]
It seems that this PR broke the "Non DB" tests for Always and CLI in https://github.com/apache/airflow/actions/runs/11200846409/job/31135117858?pr=42773 and on main.

Any idea on how to fix this (easily) w/o reversion?

**Comment by dstandish** [comment]
> It seems that this PR broke the "Non DB" tests for Always and CLI in https://github.com/apache/airflow/actions/runs/11200846409/job/31135117858?pr=42773 and on main.
> 
> Any idea on how to fix this (easily) w/o reversion?

I took a quick look but cannot reproduce locally.

I am unavailable most of today.

**Comment by jscheffl** [comment]
> > It seems that this PR broke the "Non DB" tests for Always and CLI in https://github.com/apache/airflow/actions/runs/11200846409/job/31135117858?pr=42773 and on main.
> > Any idea on how to fix this (easily) w/o reversion?
> 
> I took a quick look but cannot reproduce locally.
> 
> I am unavailable most of today.

Thanks for checking. I saw it on multiple PRs and might be it is related to non DB tests and threading?
On my Laptop I can re-produce via:
```
git checkout main
git pull
breeze ci-image build --python 3.8  # ensure this is up-to-date
breeze testing non-db-tests --parallel-test-types "Always CLI"
```

**Pull request merged**